plugins/ImageGraph/API.php

```diff
@@ -129,404 +130,453 @@ public function get(
-        }
-
-        $useUnicodeFont = array(
-            'am', 'ar', 'el', 'fa', 'fi', 'he', 'ja', 'ka', 'ko', 'te', 'th', 'zh-cn', 'zh-tw',
-        );
-        $languageLoaded = Translate::getLanguageLoaded();
-        $font = self::getFontPath(self::DEFAULT_FONT);
-        if (in_array($languageLoaded, $useUnicodeFont)) {
-            $unicodeFontPath = self::getFontPath(self::UNICODE_FONT);
-            $font = file_exists($unicodeFontPath) ? $unicodeFontPath : $font;
-        }
-
-        // save original GET to reset after processing. Important for API-in-API-call
-        $savedGET = $_GET;
-
-        try {
-            $apiParameters = array();
-            if (!empty($idGoal)) {
-                $apiParameters = array('idGoal' => $idGoal);
+        $idSiteArr = array();
+        if (strpos($idSite, ',') !== false) {
+            $idSiteList = explode(',', $idSite);
+            foreach ($idSiteList as $key) {
+                $idSiteArr[] = $key;
             }
-            // Fetch the metadata for given api-action
-            $parameters = array(
-                'idSite' => $idSite,
-                'apiModule' => $apiModule,
-                'apiAction' => $apiAction,
-                'apiParameters' => $apiParameters,
-                'language' => $languageLoaded,
-                'period' => $period,
-                'date' => $date,
-                'hideMetricsDoc' => false,
-                'showSubtableReports' => true
-            );
-
-            $metadata = Request::processRequest('API.getMetadata', $parameters);
-            if (!$metadata) {
-                throw new Exception('Invalid API Module and/or API Action');
+        } else if (!empty($idSite) && ($idSite != 'all')) {
+            $idSiteArr[] = $idSite;
+        } else if (!empty($idSite) && ($idSite == 'all')) {
+            $viewableIdSites = APISitesManager::getInstance()->getSitesIdWithAtLeastViewAccess();
+            foreach ($viewableIdSites as $site) {
+                $idSiteArr[] = $site;
             }
+        } else {
+            // empty or invalid idSite has been already handled somewhere
+            exit;
+        }
 
-            $metadata = $metadata[0];
-            $reportHasDimension = !empty($metadata['dimension']);
-            $constantRowsCount = !empty($metadata['constantRowsCount']);
-
-            $isMultiplePeriod = Period::isMultiplePeriod($date, $period);
-            if (!$reportHasDimension && !$isMultiplePeriod) {
-                throw new Exception('The graph cannot be drawn for this combination of \'date\' and \'period\' parameters.');
+        $pngImgArr = array();
+        // remove duplicated idSite; improves image graph generation time.
+        $idSiteArr = array_unique($idSiteArr);
+        $idSiteArrCount = count($idSiteArr);
+        for ($w = 0; $w < $idSiteArrCount; $w++)
+        {
+            $idSite = $idSiteArr[$w];
+            Piwik::checkUserHasViewAccess($idSite);
```

Comment:
>  "The diff is quite big indeed. I wonder if it was possible to make the code ..."

Hi Thomas, there're actually very few changes added to existing code. **Not big** as you said. Github diff tool may make it seem too big. What I did in **ImageGraph.get** function is just put a **for loop** around the statements inside the function and storing the image graph outputs to an array variable so that it could be used later. It's that simple. It doesn't alter the output behavior of any other function parameters supplied to this function. If you think that it really affects the existing behavior, then we may move it to new function. Otherwise, I don't find any reason to reuse the code and write the new functionality in a new function. Moreover, that would increase the program LOC drastically which is really not useful. Please apply this patch in you app and play around with this a few more times. You may find something more constructive. What do you think?